velox/exec/Merge.cpp

```diff
@@ -165,60 +172,224 @@ RowVectorPtr Merge::getOutput() {
+  spillSourceRows_.resize(batchSize);
+
+  VELOX_CHECK_GT(output_->size(), 0);
+  VELOX_CHECK_LE(output_->size() + numOutputRows_, numInputRows_);
+
+  int32_t outputRow = 0;
+  int32_t outputSize = 0;
+  bool isEndOfBatch = false;
+  while (outputRow + outputSize < output_->size()) {
+    SpillMergeStream* stream = spillMerger_->next();
+    VELOX_CHECK_NOT_NULL(stream);
+
+    spillSources_[outputSize] = &stream->current();
+    spillSourceRows_[outputSize] = stream->currentIndex(&isEndOfBatch);
+    ++outputSize;
+    if (FOLLY_UNLIKELY(isEndOfBatch)) {
+      // The stream is at end of input batch. Need to copy out the rows before
+      // fetching next batch in 'pop'.
+      gatherCopy(
+          output_.get(),
+          outputRow,
+          outputSize,
+          spillSources_,
+          spillSourceRows_,
+          {});
+      outputRow += outputSize;
+      outputSize = 0;
+    }
+    // Advance the stream.
+    stream->pop();
+  }
+  VELOX_CHECK_EQ(outputRow + outputSize, output_->size());
+
+  if (FOLLY_LIKELY(outputSize != 0)) {
+    gatherCopy(
+        output_.get(),
+        outputRow,
+        outputSize,
+        spillSources_,
+        spillSourceRows_,
+        {});
+  }
+
+  numOutputRows_ += output_->size();
+  return std::move(output_);
+}
+
+void MergeBuffer::maybeStartMoreSources(
+    size_t& numStartedSources,
+    std::vector<ContinueFuture>& sourceBlockingFutures,
+    uint32_t maxMergeSources,
+    vector_size_t outputBatchSize,
+    const std::vector<std::shared_ptr<MergeSource>>& sources,
+    const std::vector<SpillSortKey>& sortingKeys) {
+  if (sourceStreamMerger_ == nullptr && numStartedSources < sources.size()) {
+    VELOX_CHECK(sourceStreams_.empty());
+    // Plans merge sources for this run.
+    std::vector<MergeSource*> currentSources;
+    for (auto i = numStartedSources;
+         i < (std::min(sources.size(), numStartedSources + maxMergeSources));
+         ++i) {
+      currentSources.push_back(sources[i].get());
+    }
+
+    // Initializes the input merger.
+    std::vector<std::unique_ptr<SourceStream>> currentCursors;
+    currentCursors.reserve(currentSources.size());
+    for (auto* source : currentSources) {
+      currentCursors.push_back(
+          std::make_unique<SourceStream>(source, sortingKeys, outputBatchSize));
+    }
+
+    // mergeStreams_ and inputMerger_ are paired.
+    VELOX_CHECK(sourceStreams_.empty());
+    for (auto& cursor : currentCursors) {
+      sourceStreams_.push_back(cursor.get());
+    }
+    sourceStreamMerger_ =
+        std::make_unique<TreeOfLosers<SourceStream>>(std::move(currentCursors));
+
+    // Start sources.
+    for (const auto& source : currentSources) {
+      source->start();
+    }
+    numStartedSources += currentSources.size();
+  }
+
+  if (sourceBlockingFutures.empty()) {
+    for (const auto& cursor : sourceStreams_) {
+      cursor->isBlocked(sourceBlockingFutures);
+    }
+  }
+}
+
+void MergeBuffer::createSpillMerger(
```

Comment:
It seems that we want to separate the spill path from the input source stream path? Thanks!